test/transaction_changes_test.rb

```diff
@@ -110,4 +110,18 @@ def test_transaction_changes_type_cast
     end
     assert_empty @user.stored_transaction_changes
   end
+
+  def test_serialized_attributes_value
+    @user.connection_details = [User::ConnectionDetails.new(client_ip: '1.1.1.1')]
+    @user.save!
+    assert_instance_of Array, @user.stored_transaction_changes['connection_details']
+  end
+
+  def test_serialized_attributes_mutation
+    details = User::ConnectionDetails.new(client_ip: '1.1.1.1')
+    @user.connection_details = [details]
+    details.client_ip = '2.2.2.2'
+    @user.save!
+    assert_equal '2.2.2.2', @user.connection_details.first.client_ip
```

Comment:
This behaviour is very subtle and is kind of a gotcha within Active Record itself for application code.  Basically, this is showing that lazy casting of attribute values is a lazy abstraction for writes from application code.  As in the same problem can be seen (even without using `ArTransactionChanges`) by adding in an attribute read in the application code

```diff
     details = User::ConnectionDetails.new(client_ip: '1.1.1.1')
     @user.connection_details = [details]
+    @user.connection_details
     details.client_ip = '2.2.2.2'
     @user.save!
     assert_equal '2.2.2.2', @user.connection_details.first.client_ip
```

which would result in the same failure

```
  1) Failure:
TransactionChangesTest#test_serialized_attributes_mutation [/Users/dylants/src/ar_transaction_changes/test/transaction_changes_test.rb:126]:
Expected: "2.2.2.2"
  Actual: "1.1.1.1"
```

These mutating reads don't seem like ideal behaviour for Active Record.  It seems like it would be easier to reason about the code if any attribute value duplication happened eagerly on attribute assignment.  Of course, any attempt to fix this behaviour upstream in Active Record would be a breaking change, which would require deprecating mutations to attribute assignment arguments between the assignment and when the attribute is first read.

We should also fix any application code affected by this regression, since that code is fragile to attribute reads being introduced which could subtly prevent the serialized value mutation to not affect what is being persisted.  So if the intention is to mutate an assigned attribute value, then it should be done after reading from the attribute.  E.g.

```diff
     details = User::ConnectionDetails.new(client_ip: '1.1.1.1')
     @user.connection_details = [details]
-    details.client_ip = '2.2.2.2'
+    @user.connection_details.client_ip = '2.2.2.2'
     @user.save!
```

I'm not trying to say we shouldn't fix the regression introduced by this library, just trying to clarify the problem.

Reply:
> We should also fix any application code affected by this regression

I agree with that, but I kinda have the same issue. I can easily fix the few occurences of it in the test suite, but I'm not very confident about all the production cases being covered by test, hence why I opted to go a long way for preserving the exact behavior.